### ISSUE TITLE
[Roadmap] 1.4.0 Roadmap

### ISSUE THREAD

**Issue by trivialfis** [dmlc/xgboost#6500]
@dmlc/xgboost-committer Please add your items here by editing this post. Let's ensure that

* Each item has to be associated with a ticket
* Major design/refactoring are associated with a RFC before committing the code
* Blocking issue must be marked as blocking
* Breaking change must be marked as breaking

For other contributors who have no permission to edit the post, please comment here about what you think should be in 1.4.0.

## Main
- [x] Initial support for Python type hints. (#6496, #6519)
- [ ] Improve ranking implementation. (#6450)
- [x] Add base margin to evaluation dataset. (#6591)
- [x] Further development for inplace prediction. (#6304 #6653)
- [ ] Further development on categorical data support. (#6503)
- [ ] Add score tree interval to early stopping. (#5238, #5090)
- [x] Remove saving binary format for memory snapshot. (#6513)
- [ ] Staged predict (https://github.com/dmlc/xgboost/issues/5006)
- [x] Global configuration.  (#6414)
- [x] Training continuation with early stopping. (#6506)
- [x] Add `use_rmm` flag to global configuration (#6656)
- [x] Deprecate `ntree_limit` in Python.
- [x] Use `inplace_predict` for sklearn. (#6718)
- [x] Make all prediction functions for `gbtree` thread safe. (#6648)
- [x] Reimplement AUC metric (#6747).
- [x] Fix gamma-deviance metric. (#6731, #6761)

## Dask
- [x] Make dask scikit-learn interface feature complete. (#6471, #6602)
  + [x] Add ranking for dask. (#6576)
  + [x] Add other random forest interfaces. (#6602)
  + [x] Use inplace prediction for dask. (#6668)
- [x] Support training multiple models in parallel using dask. (#6743)
- [x] Optimize prediction performance. (#6614)
- [x] Support shap value computation. (#6645)

For brief notes, at 1.4, dask interface should be feature complete, categorical data support for GPU is ready for public testing and inplace prediction will be more mature.

**Comment by Roffild** [comment]
https://github.com/dmlc/xgboost/pull/6507

**Comment by SmirnovEgorRu** [comment]
I want to propose to replace 'approx' to 'hist when tree_method is set to 'auto' on CPU. I observed https://github.com/dmlc/xgboost/pull/5178 was about this, but it's closed.

`hist` is faster than `approx` and my observations that even accuracy is better/on par. GPUs also have only 'hist' method, not 'approx'.

Do we have any concerns not to do this?

CC: @trivialfis, @hcho3, @ShvetsKS

**Comment by trivialfis** [comment]
I will put up some documents on theoretical aspect of various tree methods, then we can decide together.

**Comment by trivialfis** [comment]
in https://github.com/dmlc/xgboost/pull/6564

**Comment by SmirnovEgorRu** [comment]
@trivialfis, do we need to run experiments to decide, probably?

**Comment by trivialfis** [comment]
@SmirnovEgorRu I don't have objection of changing the default in this or next release.  I mentioned there's a huge refactor for CPU implementations to @ShvetsKS .  I would like to see some parts of it merged before making the change so we can make some fair comparison.  Will come back to it after sorting out issues in dask interface. (which should be quite fast as most of the features are now supported).

**Comment by trivialfis** [comment]
I closed the PR you referenced because I couldn't get all tests passing, I think even if we decided to change the default now, we still have some blockers to track down.  So refactoring first might help making the change clearer and easier.

**Comment by trivialfis** [comment]
Hi @ShvetsKS @SmirnovEgorRu I have been trying to refactor the CPU code for categorical data support based on the efficient CPU Hist code.  I found the on URL dataset the cpu hist is slower than approx.  It's not a conventional dataset as it's unusually wide and sparse.  Just curious if you have plan on optimizing it.

**Comment by trivialfis** [comment]
The approx implementation is parallelizing on features with dynamic scheduling, so it has an advantage on these kind of datasets.

**Comment by SmirnovEgorRu** [comment]
@trivialfis, yep, we are thinking how to tune wide data sets as well. I suppose we can outperform `approx` with `hist` on URL.

**Comment by Denisevi4** [comment]
"Support training multiple models in parallel using dask". Does this include cross-validation with early stopping?

**Comment by trivialfis** [comment]
@Denisevi4 No, it's for running multiple training sessions on a single cluster simultaneously.  But it's a basic requirement for cv.

**Comment by Roffild** [comment]
#6731

**Comment by trivialfis** [comment]
@hcho3 I would like to get the 1.4 out once we got AUC re-implemented.  I can try fixing the gamma metric if the AUC re-implementation goes well.

**Comment by trivialfis** [comment]
I will branch out next week.

**Comment by trivialfis** [comment]
I will branch out next week.

**Comment by Roffild** [comment]
Will you fix other metrics (gamma-nloglik, logloss)?

**Comment by trivialfis** [comment]
Yeah, I will take a deeper look into them this weekend.

**Comment by trivialfis** [comment]
@Roffild Will reply on the original thread: https://github.com/dmlc/xgboost/issues/6731

**Comment by trivialfis** [comment]
1.4 is out, submit status will be on https://github.com/dmlc/xgboost/issues/6793 .
